[REEF-1327] Creating task states and task state transitions

jwang98052 opened the pull request:
* Adding Task States, task events, Task Transition
* Adding DriverTaskState
* Adding Driver test project and task cases for Driver task states transitions

JIRA: [REEF-1327](https://issues.apache.org/jira/browse/REEF-1327)

This closes #

markusweimer commented on lang/cs/Org.Apache.REEF.Driver/Task/TaskStates.cs:
```diff
@@ -0,0 +1,52 @@
+    /// <summary>
+    /// Task states in task transitions
+    /// </summary>
+    public enum TaskStates
```
Maybe call this `TaskState` and `TaskEvent`? It makes for an easier read where it is used. Right now, those cases read as if they are collections because of the plural form.

jwang98052 replied:
Done

jwang98052 replied:
TaskState is used in Evaluator namespace. It is easy to cause ambiguous if the same name is used. I changed it to TaskTransitionState.

markusweimer commented on lang/cs/Org.Apache.REEF.Driver/Task/TaskStates.cs:
```diff
@@ -0,0 +1,52 @@
+    /// </summary>
+    public enum TaskEvents
+    {
+        SubmittingTask,
```
The names here don't describe events. For instance, please replace `SubmittingTask` with `Submitted`, as that describes a precise time: right after the task is submitted.

markusweimer commented on lang/cs/Org.Apache.REEF.Driver/Task/TaskStates.cs:
```diff
@@ -0,0 +1,52 @@
+        FailTaskAppError,
+        FailTaskSystemError,
+        FaileTaskEvaluatorError,
+        FailTaskCommuError
```
Please don't use abbreviations here. IntelliSense will ease the burden of typing the full words :)

jwang98052 replied:
Done

markusweimer commented on lang/cs/Org.Apache.REEF.Driver/Task/DriverTaskState.cs:
```diff
@@ -0,0 +1,88 @@
+
+namespace Org.Apache.REEF.Driver.Task
+{
+    public class DriverTaskState
```
Please make it `sealed` and add documentation.

Also: What happens if the event received is not a possible event in the current state? We should generate an exception in those cases.

jwang98052 replied:
Done. 
Yes, it throws Exception in GetNext()
 throw new Exception("Invalid transition: " + _currentState + " -> " + taskEvent);

jwang98052 replied:
I have test cases for not allowed transitions.

markusweimer commented on lang/cs/Org.Apache.REEF.Driver/Task/DriverTaskState.cs:
```diff
@@ -0,0 +1,88 @@
+            }
+        }
+
+        public TaskStates GetNext(TaskEvents taskEvent)
```
This doesn't have to be `public` right?

jwang98052 replied:
Done

markusweimer commented:
I've done a pass.

jwang98052 commented:
@markusweimer I have addressed your review comments.

afchung commented on lang/cs/Org.Apache.REEF.Driver.Tests/Org.Apache.REEF.Driver.Tests.csproj:
```diff
@@ -0,0 +1,70 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Why do we need a new test project?

jwang98052 replied:
This is for driver unit testing. We usually have unit tests for each project if needed. REEF.Tests is for E2E testing. 

Currently there is not unit test project for driver.

afchung commented on lang/cs/Org.Apache.REEF.Driver.Tests/TestTaskStates.cs:
```diff
@@ -0,0 +1,218 @@
+            {
+                Assert.True(e.Message.Contains("Invalid transition:"));
+            }
+
```
You also need to validate that an `Exception` is actually thrown.

jwang98052 replied:
done

afchung commented on lang/cs/Org.Apache.REEF.Driver/Task/DriverTaskState.cs:
```diff
@@ -0,0 +1,88 @@
+
+namespace Org.Apache.REEF.Driver.Task
+{
+    sealed public class DriverTaskState
```
`public sealed`.

afchung replied:
Document.

jwang98052 replied:
done

afchung commented on lang/cs/Org.Apache.REEF.Driver/Task/DriverTaskState.cs:
```diff
@@ -0,0 +1,88 @@
+        {
+            get
+            {
+                lock (_stateLock)
```
Do we really need this lock? We can just use `volatile` right?

jwang98052 replied:
Ok

afchung commented on lang/cs/Org.Apache.REEF.Driver/Task/TaskStates.cs:
```diff
@@ -0,0 +1,52 @@
+    /// <summary>
+    /// Task events that triggers task state transition
+    /// </summary>
+    public enum TaskEvent
```
Split into a separate file.

jwang98052 replied:
done

afchung commented on lang/cs/Org.Apache.REEF.Driver/Task/TaskStates.cs:
```diff
@@ -0,0 +1,52 @@
+    /// </summary>
+    public enum TaskState
+    {
+        TaskNew,
```
Could you add XML docs to each state? Some of them aren't entirely intuitive.

jwang98052 replied:
Done

afchung commented on lang/cs/Org.Apache.REEF.Driver/Task/TaskStates.cs:
```diff
@@ -0,0 +1,52 @@
+    /// </summary>
+    public enum TaskEvent
+    {
+        SubmittedTask,
```
XML docs.

jwang98052 replied:
done

afchung commented on lang/cs/Org.Apache.REEF.Driver.Tests/TestTaskStates.cs:
```diff
@@ -0,0 +1,218 @@
+            {
+                Assert.True(e.Message.Contains("Invalid transition:"));
+            }
+
```
Assert `Exception` caught.

jwang98052 replied:
updated

afchung commented:
Done with a first pass.

jwang98052 commented:
@afchung I have addressed your comments.

afchung commented on lang/cs/Org.Apache.REEF.Driver/Task/TaskTransitionState.cs:
```diff
@@ -0,0 +1,75 @@
+        /// <summary>
+        /// After receiving CompletedTask event, the task state is TaskCompeleted
+        /// </summary>
+        TaskCompeleted,
```
Typo

jwang98052 replied:
done

afchung commented on lang/cs/Org.Apache.REEF.Driver/Task/TaskStateTransition.cs:
```diff
@@ -0,0 +1,45 @@
+    /// <summary>
+    /// It defines task state and task event combination that would determine the next transition state 
+    /// </summary>
+    public class TaskStateTransition
```
`sealed`.

jwang98052 replied:
done

afchung commented on lang/cs/Org.Apache.REEF.Driver/Task/DriverTaskState.cs:
```diff
@@ -0,0 +1,84 @@
+
+        private TaskTransitionState GetNext(TaskEvent taskEvent)
+        {
+            TaskStateTransition transition = new TaskStateTransition(_currentState, taskEvent);
```
Don't use `_currentState` here, if you would like to keep `_currentState` as `volatile` without using the lock. Instead, pass in `_currentState` with the function call and make it static.

afchung replied:
The alternative is to switch back to `lock`.

jwang98052 replied:
I might just use lock as _currentState is already an instance variable.

afchung commented on lang/cs/Org.Apache.REEF.Driver/Task/DriverTaskState.cs:
```diff
@@ -0,0 +1,84 @@
+    /// <summary>
+    /// DriverTaskState wraps current state and provides methods to move from one state to the next state
+    /// </summary>
+    sealed public class DriverTaskState
```
`public sealed`.

jwang98052 replied:
done

afchung commented on lang/cs/Org.Apache.REEF.Driver.Tests/TestTaskStates.cs:
```diff
@@ -0,0 +1,146 @@
+    public class TestTaskStates
+    {
+        [Fact]
+        public void TestHappySenario()
```
Rename this. What is this the Happy Scenario of?

jwang98052 replied:
done

afchung commented:
@jwang98052 done with a second pass.

jwang98052 commented:
@afchung addressed review comments.

afchung commented on lang/cs/Org.Apache.REEF.Driver/Task/DriverTaskState.cs:
```diff
@@ -0,0 +1,88 @@
+            { new TaskStateTransition(TaskTransitionState.TaskFailedByGroupCommunication, TaskEvent.FailedTaskEvaluatorError), TaskTransitionState.TaskFailedByEvaluatorFailure }
+        };
+
+        volatile private TaskTransitionState _currentState;
```
Nit: `private volatile`.

jwang98052 replied:
done

afchung commented on lang/cs/Org.Apache.REEF.Driver/Task/DriverTaskState.cs:
```diff
@@ -0,0 +1,88 @@
+        };
+
+        volatile private TaskTransitionState _currentState;
+        readonly private object _lock = new object();
```
Nit: `private readonly`.

jwang98052 replied:
done

afchung commented:
@jwang98052 final 2 nits, and then I'm signed off. I'll leave this to @markusweimer to sign off as well. Thanks!

jwang98052 commented:
@afchung update checked in.

tcNickolas commented:
Several notes related to creating new test project.
1. Why is there no packages.config? All other test projects have their own, even if packages declared in them duplicate the ones imported by other projects.
2. With the current state of AppVeyor config, we need to list each test project so that it gets executed. Please add a line to appveyor.yml similar to other projects, so that we can verify there's no issue with these tests manifests in AppVeyor only.

tcNickolas commented on lang/cs/Org.Apache.REEF.Driver.Tests/TestTaskStates.cs:
```diff
@@ -0,0 +1,146 @@
+        public void TestNewToCompleteSenario()
+        {
+            var taskState = new DriverTaskState();
+            Assert.True(taskState.CurrentState.Equals(TaskTransitionState.TaskNew));
```
I dislike using Assert.True without an message a bit, because when AppVeyor test fails, it only shows "Expected True, Actual False" and line number in stack trace. I feel it's nice to have a meaningful message in Assert. Do you think it's worth adding messages to these Asserts?

jwang98052 replied:
Done

jwang98052 replied:
Added tests to appveyor.yml

jwang98052 replied:
Usually packages.config is added when we first reference some NuGet. For now, we don't reference any NuGet yet.

tcNickolas replied:
We reference StyleCop (in all projects) and xUnit NuGets (in all test projects). So I think we do need packages.config.

tcNickolas commented on lang/cs/Org.Apache.REEF.Driver.Tests/TestTaskStates.cs:
```diff
@@ -0,0 +1,146 @@
+        public void TestRunningToCloseSenario()
+        {
+            var taskState = new DriverTaskState();
+            Assert.True(taskState.CurrentState.Equals(TaskTransitionState.TaskNew));
```
A lot of Asserts seem excessive to me. For example, transition new-submitted-running is covered by 4 tests. Maybe we should verify this sequence in first test only, and in the next ones we should verify only steps which are unique to these tests?

jwang98052 replied:
Done

tcNickolas commented on lang/cs/Org.Apache.REEF.Driver/Task/DriverTaskState.cs:
```diff
@@ -0,0 +1,88 @@
+    /// <summary>
+    /// DriverTaskState wraps current state and provides methods to move from one state to the next state
+    /// </summary>
+    public sealed class DriverTaskState
```
Based on our design document, we should have a function that encapsulates logic of "whether task state is final state".

jwang98052 replied:
Added.

afchung replied:
`DriverTaskStateMachine`.

jwang98052 commented:
@tcNickolas I have addressed your review comments.

tcNickolas commented on lang/cs/Org.Apache.REEF.Driver.Tests/TestTaskStates.cs:
```diff
@@ -0,0 +1,152 @@
+        public void TestFromNewToNotAllowed()
+        {
+            var taskState = new DriverTaskState();
+            Assert.True(taskState.CurrentState.Equals(TaskTransitionState.TaskNew));
```
This case is covered in `TestNewToCompleteSenario` test

jwang98052 replied:
Done

tcNickolas commented on lang/cs/Org.Apache.REEF.Driver.Tests/TestTaskStates.cs:
```diff
@@ -0,0 +1,152 @@
+        public void TestFromRunningToNotAllowed()
+        {
+            var taskState = new DriverTaskState();
+            Assert.True(taskState.CurrentState.Equals(TaskTransitionState.TaskNew), "Task initial state is not TaskNew.");
```
This case is covered in TestNewToCompleteSenario test, don't need assert here

jwang98052 replied:
Done

tcNickolas commented on lang/cs/Org.Apache.REEF.Driver.Tests/TestTaskStates.cs:
```diff
@@ -0,0 +1,152 @@
+            Assert.True(taskState.CurrentState.Equals(TaskTransitionState.TaskNew), "Task initial state is not TaskNew.");
+
+            taskState.MoveNext(TaskEvent.SubmittedTask);
+            Assert.True(taskState.CurrentState.Equals(TaskTransitionState.TaskSubmitting), "Fail to move to TaskSubmitting state.");
```
This case is covered in TestNewToCompleteSenario test, don't need assert here (and the next assert also)

jwang98052 replied:
Done

tcNickolas commented on lang/cs/Org.Apache.REEF.Driver.Tests/TestTaskStates.cs:
```diff
@@ -0,0 +1,152 @@
+        public void TestFromFailToNotAllowed()
+        {
+            var taskState = new DriverTaskState();
+            Assert.True(taskState.CurrentState.Equals(TaskTransitionState.TaskNew), "Task initial state is not TaskNew.");
```
This case is covered in TestNewToCompleteSenario test, don't need assert here (and two next asserts)

jwang98052 replied:
Done

tcNickolas commented on lang/cs/Org.Apache.REEF.Driver.Tests/TestTaskStates.cs:
```diff
@@ -0,0 +1,152 @@
+        {
+            var taskState = new DriverTaskState();
+            Assert.True(taskState.CurrentState.Equals(TaskTransitionState.TaskNew));
+            Assert.False(taskState.IsFinalState(), "TaskNew is not final state.");
```
Error message should be "TaskNew should not be a final state" (if Assert fails, then it is a final state, and error message says it is not, which is confusing). Same for the rest of `Assert.False(taskState.IsFinalState())` checks

jwang98052 replied:
done

dkm2110 commented on lang/cs/Org.Apache.REEF.Driver/Task/TestEvent.cs:
```diff
@@ -0,0 +1,70 @@
+        /// <summary>
+        /// Received failed Evaluator that caused associated task fail
+        /// </summary>
+        FailedTaskEvaluatorError,
```
What is the need for this state?  Isn't FailedEvaluator more relevant in this stage?

jwang98052 replied:
This is a state for task. When we receive a FailedEvaluator event, we need to set a state for associated task, showing that the task failure is caused by evaluator error. Please refer to the design doc.

dkm2110 commented on lang/cs/Org.Apache.REEF.Driver/Task/TestEvent.cs:
```diff
@@ -0,0 +1,70 @@
+        /// <summary>
+        /// Received failed task with communication error
+        /// </summary>
+        FailedTaskCommunicationError
```
Communication with what? Driver, other evaluators or either?

jwang98052 replied:
This is the failure caused by group communication between tasks. Ref design doc:
"received failed task event with the message showing the task failure is caused by not able to get messages in the communication group"
At task side, we need to throw proper exception with proper error message to represent such communication error. I will create a Jira for it,

dkm2110 commented on lang/cs/Org.Apache.REEF.Driver/Task/TestEvent.cs:
```diff
@@ -0,0 +1,70 @@
+        /// <summary>
+        /// Received failed task with system error
+        /// </summary>
+        FailedTaskSystemError,
```
Can you give me a system error example not covered by FailedTaskCommunicationError or FailedTaskEvaluatorError

jwang98052 replied:
We have REEF-1225 to identify such  cases. I am not sure if FailedTaskCommunicationError and FailedTaskEvaluatorError can represent all the system errors. We introduce it in the design to cover those that are not belong to those two cases.

dkm2110 replied:
I see .... but doesn't these make things complicated with two more states needing to be tracked? Especially given currently we cannot think of any examples for these.

tcNickolas replied:
Out of memory - it's not communication error and not app error.

dkm2110 commented on lang/cs/Org.Apache.REEF.Driver/Task/TaskTransitionState.cs:
```diff
@@ -0,0 +1,75 @@
+        TaskFailedByEvaluatorFailure,
+
+        /// <summary>
+        /// After receiving FaiedTask event and verified it is caused by group communication
```
Typo

jwang98052 replied:
Done

tcNickolas commented on lang/cs/Org.Apache.REEF.Driver/Task/DriverTaskState.cs:
```diff
@@ -0,0 +1,129 @@
+            { new TaskStateTransition(TaskTransitionState.TaskFailedByGroupCommunication, TaskEvent.FailedTaskEvaluatorError), TaskTransitionState.TaskFailedByEvaluatorFailure }
+        };
+
+        private static TaskTransitionState[] FinalStatae = 
```
Typo: FinalStates

afchung replied:
`readonly`. Why not use `Set`?

markusweimer replied:
+1 on using `Set`, preferably an immutable one.

jwang98052 replied:
Changed to ISet. From the usage, it is a static set build one time only. I think ISet should be fine.

dkm2110 commented on lang/cs/Org.Apache.REEF.Driver/Task/TaskTransitionState.cs:
```diff
@@ -0,0 +1,75 @@
+        TaskFailedByGroupCommunication,
+
+        /// <summary>
+        /// After receiving FaiedTask event and verified it is caused by application error
```
Typo

jwang98052 replied:
done

dkm2110 commented on lang/cs/Org.Apache.REEF.Driver/Task/TaskTransitionState.cs:
```diff
@@ -0,0 +1,75 @@
+        TaskFailedAppError,
+
+        /// <summary>
+        /// After receiving FaiedTask event and verified it is caused by system error
```
typo

jwang98052 replied:
done

tcNickolas commented on lang/cs/Org.Apache.REEF.Driver/Task/DriverTaskState.cs:
```diff
@@ -0,0 +1,129 @@
+        /// <returns></returns>
+        public bool IsFinalState()
+        {
+            return FinalStatae.Contains(_currentState);
```
I'd use `IsFinalState(_currentState)` here (so that the logic of defining final state is in only one place, not two)

dkm2110 replied:
typo

jwang98052 replied:
done

dkm2110 commented on lang/cs/Org.Apache.REEF.Driver/Task/DriverTaskState.cs:
```diff
@@ -0,0 +1,129 @@
+namespace Org.Apache.REEF.Driver.Task
+{
+    /// <summary>
+    /// DriverTaskState wraps current state and provides methods to move from one state to the next state
```
I would introduce some good serious documentation here rather than one line. The person looking it at first time will have all kind of doubts.

afchung replied:
Or provide a link to the design doc or some sort of diagram.

jwang98052 replied:
done

afchung commented on lang/cs/Org.Apache.REEF.Driver/Task/TaskTransitionState.cs:
```diff
@@ -0,0 +1,75 @@
+    /// <summary>
+    /// Task states in task transitions
+    /// </summary>
+    public enum TaskTransitionState
```
Why not just `TaskState` or `DriverTaskState`?

markusweimer replied:
+1 on `DriverTaskState`

jwang98052 replied:
DriverTaskState has been used for the class that represents task state and transitions. 
I was using TaskState, but it is easy to cause ambiguous with the Org.Apache.REEF.Common.Runtime.Evaluator.Task.TaskState. May be I just use this name  back as they are in different name spaces and the usages are only in a few classes.

dkm2110 commented on lang/cs/Org.Apache.REEF.Driver/Task/DriverTaskState.cs:
```diff
@@ -0,0 +1,129 @@
+    /// </summary>
+    public sealed class DriverTaskState
+    {
+        internal static IDictionary<TaskStateTransition, TaskTransitionState> Transitions = new Dictionary<TaskStateTransition, TaskTransitionState>
```
So if I understnad correctly these are valid transitions that are allowed?

markusweimer replied:
Yes, I think so. However, why is this `internal` and not `private`?

jwang98052 replied:
Private is Ok.

dkm2110 commented on lang/cs/Org.Apache.REEF.Driver/Task/TaskTransitionState.cs:
```diff
@@ -0,0 +1,75 @@
+        /// <summary>
+        /// After submit a task with an IActiveContext, the task state is changed to TaskSubmitting
+        /// </summary>
+        TaskSubmitting,
```
Isn't TaskSubmitted a better word?

markusweimer replied:
Agreed, `TaskSubmitted` describes the state better.

jwang98052 replied:
done

tcNickolas commented:
I've done a pass.

dkm2110 commented on lang/cs/Org.Apache.REEF.Driver/Task/DriverTaskState.cs:
```diff
@@ -0,0 +1,129 @@
+            { new TaskStateTransition(TaskTransitionState.TaskRunning, TaskEvent.FailedTaskEvaluatorError), TaskTransitionState.TaskFailedByEvaluatorFailure },
+            { new TaskStateTransition(TaskTransitionState.TaskRunning, TaskEvent.FailedTaskCommunicationError), TaskTransitionState.TaskFailedByGroupCommunication },
+            { new TaskStateTransition(TaskTransitionState.TaskWaitingForClose, TaskEvent.ClosedTask), TaskTransitionState.TaskClosedByDriver },
+            { new TaskStateTransition(TaskTransitionState.TaskWaitingForClose, TaskEvent.FailedTaskAppError), TaskTransitionState.TaskFailedAppError },
```
I have a question/comment here: while the task gets closed when driver ordered it, it might be in some intermediate state and user might not think of logic to shut down task properly in these intermediate states...so u might get recoverable group comm. or system or app. error at this stage...So I am a little confused what to do here

tcNickolas replied:
Our design doc says that if `FailedTask` handler is received and task is in `TaskWaitingForClose` state, it transitions to `TaskClosedByDriver` regardless of failure cause. Need to fix either the code or the design doc (probably a fix in the code is better).

dkm2110 replied:
Design doc is right..Code here should be modified.

jwang98052 replied:
The design doc says "If the task state is TASK_WAITING_FOR_CLOSE, update the task status as TASK_CLOSED_BY_DRIVER". But the diagram shows transit to corresponding state based on the error. 

I agree to move to TASK_CLOSED_BY_DRIVER state regardless what task error it is during the closing. The major difference is that if is closed by driver, it is still recoverable. If it is app error, it will be not recoverable. 

I will update both diagram and code if no objection.

dkm2110 commented on lang/cs/Org.Apache.REEF.Driver/Task/DriverTaskState.cs:
```diff
@@ -0,0 +1,129 @@
+        /// <returns></returns>
+        public static bool IsFinalState(TaskTransitionState taskState)
+        {
+            return FinalStatae.Contains(taskState);
```
Typo

jwang98052 replied:
done

dkm2110 commented on lang/cs/Org.Apache.REEF.Driver/Task/DriverTaskState.cs:
```diff
@@ -0,0 +1,129 @@
+        {
+            lock (_lock)
+            {
+                _currentState = GetNext(command);
```
Remove GetNext() and put code here...The current code here is 2 lines only.

markusweimer replied:
Depends. Do we want to be able to ask `what-if` on this object? `GetNext()` allows that without changing state of the object.

jwang98052 replied:
That is right. GetNext() was public. As I don't have external usage yet, so changed it to private. I would change it back to public for client may want to know the next state without really moving it.

dkm2110 commented:
@jwang98052  I left some comments...In general many typos....I might not have captured all...Some sort of spell checker might be good.

markusweimer commented on lang/cs/Org.Apache.REEF.Driver/Task/DriverTaskState.cs:
```diff
@@ -0,0 +1,129 @@
+            TaskTransitionState nextState;
+            if (!Transitions.TryGetValue(transition, out nextState))
+            {
+                throw new ApplicationException("Invalid transition: " + _currentState + " -> " + taskEvent);
```
We might want to add a specific exception type for this, which gets the `_currentState` and the `taskEvent` as parameters to the constructor and generates the `Message()` property automatically.

markusweimer commented on lang/cs/Org.Apache.REEF.Driver/Task/DriverTaskState.cs:
```diff
@@ -0,0 +1,129 @@
+        /// </summary>
+        /// <param name="command"></param>
+        /// <returns></returns>
+        public TaskTransitionState MoveNext(TaskEvent command)
```
Please document the exception behavior here.

jwang98052 replied:
done

markusweimer commented on lang/cs/Org.Apache.REEF.Driver/Task/TaskTransitionState.cs:
```diff
@@ -0,0 +1,75 @@
+        /// <summary>
+        /// After receiving FaiedTask event and verified it is closed by driver, the state is set to TaskClosedByDriver
+        /// </summary>
+        TaskClosedByDriver,
```
This doesn't sound right. If the Task has failed, why should this be `TaskClosedByDriver` state? Wouldn't it be `TaskFailed` or `TaskFailureConfirmed`?

tcNickolas replied:
Based on our design doc, in case of failure of one task or evaluator driver sends message to all other running tasks to shut them down. We expect the result of this message to be FailedTask event but it is not caused by failure of app itself (from which we don't want to recover). So we want to distinguish this scenario from task failure not ordered by driver.

markusweimer replied:
Ah, this makes sense. And it gives weight to the suggestion of moving all of this into the IMRU project and making it `internal`, as these task states can only be understood in that context.

markusweimer commented on lang/cs/Org.Apache.REEF.Driver/Task/TaskTransitionState.cs:
```diff
@@ -0,0 +1,75 @@
+        TaskWaitingForClose,
+
+        /// <summary>
+        /// After receiving FaiedTask event and verified it is closed by driver, the state is set to TaskClosedByDriver
```
Typo.

jwang98052 replied:
Done

markusweimer commented:
I've also done a pass. I agree with @dkm2110, we should apply a spell checker before merging. 

Code wise, I have another question: All the new types introduced here are `public` and in the `Driver` project. Given that the purpose of these new types is to model the state transitions in the IMRU Driver, I wonder whether they shouldn't be in the `IMRU` project and marked `internal` to keep our API surface small.

Thoughts?

jwang98052 commented:
If we want to make those classes internal before it is mature, we can put it inside IMRU project. If eventually they will be public as they are basically generic, we can leave them in this name space. 
I think TaskClosedByDriver is still generic. What is not generic is TaskFailedByGroupCommunication.

jwang98052 commented:
@markusweimer @afchung @tcNickolas @dkm2110 I have addressed your review comments.

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/DriverTaskState.cs:
```diff
@@ -0,0 +1,137 @@
+        private readonly static IDictionary<TaskStateTransition, TaskState> Transitions = new ReadOnlyDictionary<TaskStateTransition, TaskState>(
+            new Dictionary<TaskStateTransition, TaskState>
+        {
+            { new TaskStateTransition(TaskState.TaskNew, TaskEvent.SubmittedTask), TaskState.TaskSubmitted },
```
I'm noticing that the `TaskEvent` basically parallels the `TaskState`. Do we really need a full blown `StateMachine` here? Or do we just need some switch statements?

dkm2110 replied:
That is my observation too.

jwang98052 replied:
I started from switch statement and realized it is mess in the code. Using state machine is more clear for transition rules and move the code logic to data. When more states are added and more rules are added, we just need to change data instead of code.

dkm2110 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/ResultHandler/WriteResultHandler.cs:
```diff
@@ -83,7 +82,7 @@ public void Dispose()
             if (_fileSystem.Exists(remoteUri))
             {
-                Exceptions.Throw(
+                Utilities.Diagnostics.Exceptions.Throw(
```
Why is this changed as part of this pull request?

jwang98052 replied:
I added Exception in this project, it is ambiguous with the Exception in the utility project. I have to add name space to avoid conflict otherwise it causes build error.

dkm2110 replied:
Can you exactly point out the ambiguity with which other function?

jwang98052 replied:
I added Org.Apache.REEF.IMRU.OnREEF.Exceptions.TaskStateTransitionException based on the review comment. After that, all Exceptions.Throw which actually refer to Utilities.Diagnostics. shows red color in the code and cannot compile. 

I am pretty sure we will need to add more exceptions in Org.Apache.REEF.IMRU.OnREEF.Exceptions when we move on for Fault tolerant work. We have to specify name space some where to distinct those two. 

This happens already in many other projects in REEF. An example is REEF.Common.Exceptions. In other classes in REEF.Common where Utilities.Diagnostics.Exception,  we have to use namespace. 

Example is EvaluatorRuntime.cs, you can find a lot of usage like:
Utilities.Diagnostics.Exceptions.CaughtAndThrow

dkm2110 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/MapInputWithControlMessage/MapInputWithControlMessageCodec.cs:
```diff
@@ -64,7 +63,7 @@ private MapInputWithControlMessageCodec(IStreamingCodec<TMapInput> baseCodec)
             }
 
-            Exceptions.Throw(new Exception("Control message type not valid in Codec read"), Logger);
+            Utilities.Diagnostics.Exceptions.Throw(new Exception("Control message type not valid in Codec read"), Logger);
```
same as below

jwang98052 replied:
same as below

dkm2110 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/MapInputWithControlMessage/MapInputWithControlMessageCodec.cs:
```diff
@@ -112,7 +111,7 @@ private MapInputWithControlMessageCodec(IStreamingCodec<TMapInput> baseCodec)
             }
 
-            Exceptions.Throw(new Exception("Control message type not valis in Codec read"), Logger);
+            Utilities.Diagnostics.Exceptions.Throw(new Exception("Control message type not valis in Codec read"), Logger);
```
this should not change as part of this pull request

jwang98052 replied:
same as below

dkm2110 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/TestEvent.cs:
```diff
@@ -0,0 +1,70 @@
+﻿// Licensed to the Apache Software Foundation (ASF) under one
```
The class name should be "TaskEvent" instead of "TestEvent".

jwang98052 replied:
Yes, typed wrong file name. Fixed.

dkm2110 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/TestEvent.cs:
```diff
@@ -0,0 +1,70 @@
+        FailedTaskSystemError,
+
+        /// <summary>
+        /// Received failed Evaluator that caused associated task fail
```
change fail -> failure

jwang98052 replied:
done

dkm2110 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/TaskStateTransition.cs:
```diff
@@ -0,0 +1,45 @@
+
+        public override int GetHashCode()
+        {
+            return 17 + (31 * _currentState.GetHashCode()) + (31 * _taskEvent.GetHashCode());
```
What are the uniqueness guarantees associates with GetHashCode() here.....specifically with the use of 17 and 31?

jwang98052 replied:
In terms of uniqueness and even distributions, they have to be prime numbers.  I got this recommendation from web site.  Do you see any issue here?

markusweimer replied:
Aren't these methods usually IDE-generated?

dkm2110 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/TaskStateTransition.cs:
```diff
@@ -0,0 +1,45 @@
+            return other != null && _currentState == other._currentState && _taskEvent == other._taskEvent;
+        }
+    }
+}
```
Documentation is missing for all public and internal members here.

jwang98052 replied:
done

dkm2110 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/TaskState.cs:
```diff
@@ -0,0 +1,75 @@
+        /// <summary>
+        /// After receiving FailedTask event and verified it is caused by application error
+        /// </summary>
+        TaskFailedAppError,
```
ByAppError?

jwang98052 replied:
Done

dkm2110 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/TaskState.cs:
```diff
@@ -0,0 +1,75 @@
+        /// <summary>
+        /// After receiving FailedTask event and verified it is caused by system error
+        /// </summary>
+        TaskFailedSystemError
```
BySystemError?

jwang98052 replied:
done

dkm2110 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/DriverTaskState.cs:
```diff
@@ -0,0 +1,137 @@
+            { new TaskStateTransition(TaskState.TaskRunning, TaskEvent.FailedTaskSystemError), TaskState.TaskFailedSystemError },
+            { new TaskStateTransition(TaskState.TaskRunning, TaskEvent.FailedTaskEvaluatorError), TaskState.TaskFailedByEvaluatorFailure },
+            { new TaskStateTransition(TaskState.TaskRunning, TaskEvent.FailedTaskCommunicationError), TaskState.TaskFailedByGroupCommunication },
+            { new TaskStateTransition(TaskState.TaskWaitingForClose, TaskEvent.ClosedTask), TaskState.TaskClosedByDriver },
```
Where are the states here for failure events...didn't we agree to transition to TaskEvent.ClosedTask in failure cases if state is TaskState.TaskWaitingForClose

jwang98052 replied:
You are right. Updates.

dkm2110 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/DriverTaskState.cs:
```diff
@@ -0,0 +1,137 @@
+        }
+
+        /// <summary>
+        /// Checks if the current state is a final state
```
Change name to IsCurrentStateFinal.

dkm2110 replied:
You do not need this function. You can simply ask IsFinalState(CurrentState)

jwang98052 replied:
I have replied earlier for the same review comment above, in the code, most of the usage is that we have DriverTaskState object, if calling IsFinalState(CurrentState), we have to get current state then call it. Calling the first one is straightforward for a given DriverTaskState  object.

dkm2110 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs:
```diff
@@ -75,7 +74,7 @@ internal void EvaluatorFailed(string evaluatorId)
                 if (!_submittedEvaluators.Contains(evaluatorId))
                 {
-                    Exceptions.Throw(new Exception("Failed evaluator was never submitted"), Logger);
+                    Utilities.Diagnostics.Exceptions.Throw(new Exception("Failed evaluator was never submitted"), Logger);
```
All these should not be part of this pull request.

jwang98052 replied:
I get compile error without the change with the code in this PR.

dkm2110 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs:
```diff
@@ -95,12 +94,12 @@ internal ContextAndServiceConfiguration GetNextConfiguration(string evaluatorId)
                 if (_submittedEvaluators.Contains(evaluatorId))
                 {
-                    Exceptions.Throw(new Exception("The evaluator is already submitted"), Logger);
+                    Utilities.Diagnostics.Exceptions.Throw(new Exception("The evaluator is already submitted"), Logger);
```
same as above

jwang98052 replied:
same as above

dkm2110 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs:
```diff
@@ -95,12 +94,12 @@ internal ContextAndServiceConfiguration GetNextConfiguration(string evaluatorId)
                 if (_failedEvaluators.Count == 0 && _assignedPartitionDescriptors >= _numNodes)
                 {
-                    Exceptions.Throw(new Exception("No more configuration can be provided"), Logger);
+                    Utilities.Diagnostics.Exceptions.Throw(new Exception("No more configuration can be provided"), Logger);
```
same as above

jwang98052 replied:
same as above

dkm2110 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ServiceAndContextConfigurationProvider.cs:
```diff
@@ -118,7 +117,7 @@ internal ContextAndServiceConfiguration GetNextConfiguration(string evaluatorId)
                     if (_configurationProvider.ContainsKey(evaluatorId))
                     {
-                        Exceptions.Throw(
+                        Utilities.Diagnostics.Exceptions.Throw(
```
same as above

jwang98052 replied:
same as above

dkm2110 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/Driver/ConfigurationManager.cs:
```diff
@@ -59,7 +58,7 @@ internal sealed class ConfigurationManager
             catch (Exception e)
             {
-                Exceptions.Throw(e, "Unable to deserialize map function configuration", Logger);
+                Utilities.Diagnostics.Exceptions.Throw(e, "Unable to deserialize map function configuration", Logger);
```
Should not be part of this pull request. Please fix it at every place.

jwang98052 replied:
same as above

dkm2110 commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/DriverTaskState.cs:
```diff
@@ -0,0 +1,137 @@
+    /// For the task state transition diagram <see href="https://issues.apache.org/jira/browse/REEF-1327"></see>
+    /// This class wraps current state and provides methods to move from one state to the next state
+    /// </summary>
+    internal sealed class DriverTaskState
```
DriverTaskStateMachine?

jwang98052 replied:
The usage of this class most possibly will be a column in task dictionary, which represents the current task state. Calling it as a DriverTaskStateMachine seems wired. 

We can look at it again when we start to use it.

dkm2110 commented on lang/cs/Org.Apache.REEF.Evaluator.Tests/TaskRuntimeTests.cs:
```diff
@@ -63,7 +63,7 @@ public void TestTaskRuntimeInitialization()
             var injector = GetInjector();
             var taskRuntime = injector.GetInstance<TaskRuntime>();
-            Assert.Equal(taskRuntime.GetTaskState(), TaskState.Init);
+            Assert.Equal(taskRuntime.GetTaskState(), EvaluatorTaskState.Init);
```
Why are these changes coming here?

jwang98052 replied:
TaskState is defined in both IMRU and Org.Apache.REEF.Common.Runtime.Evaluator.Task. To avoid ambiguous, I defined an alias for Org.Apache.REEF.Common.Runtime.Evaluator.Task.TaskState in this class.

dkm2110 commented on lang/cs/Org.Apache.REEF.IMRU.Tests/TestTaskStates.cs:
```diff
@@ -0,0 +1,166 @@
+    /// <summary>
+    /// The test cases in this classes test TaskState and transitions
+    /// </summary>
+    public class TestTaskStates
```
Please provide detailed documentation with separate test methods

jwang98052 replied:
done

dkm2110 commented:
@jwang98052  I have taken a pass. There are lots of non related changes in this pull request.

jwang98052 commented:
@afchung @dkm2110 I have addressed the review comments.

@dkm2110 All the changed classes are related to this PR, not logically but from build perspective as there are some name conflict.

markusweimer commented on appveyor.yml:
```diff
@@ -42,6 +42,7 @@ test_script:
   - cmd: .\packages\xunit.runner.console.2.1.0\tools\xunit.console.exe .\bin\x64\Debug\Org.Apache.REEF.IMRU.Tests\Org.Apache.REEF.IMRU.Tests.dll
   - cmd: .\packages\xunit.runner.console.2.1.0\tools\xunit.console.exe .\bin\x64\Debug\Org.Apache.REEF.IO.Tests\Org.Apache.REEF.IO.Tests.dll
   - cmd: .\packages\xunit.runner.console.2.1.0\tools\xunit.console.exe .\bin\x64\Debug\Org.Apache.REEF.Network.Tests\Org.Apache.REEF.Network.Tests.dll
+  - cmd: .\packages\xunit.runner.console.2.1.0\tools\xunit.console.exe .\bin\x64\Debug\Org.Apache.REEF.Driver.Tests\Org.Apache.REEF.Driver.Tests.dll
```
We don't need this anymore, right?

jwang98052 replied:
That's right.

jwang98052 commented:
@markusweimer I have addressed your comments.

afchung commented on lang/cs/Org.Apache.REEF.IMRU/OnREEF/IMRUTasks/TaskStateTransition.cs:
```diff
@@ -0,0 +1,61 @@
+    /// <summary>
+    /// It defines task state and task event combination that would determine the next transition state 
+    /// </summary>
+    internal sealed class TaskStateTransition
```
Please remove this after #948 is merged.

jwang98052 replied:
Please merge #948 first and I will remove the redundant class and move the classes into StateMachine if that is a better location.

afchung commented:
@markusweimer @dkm2110 @tcNickolas this should only be merged after #948, since there are redundant classes, or @jwang98052 may want to close this PR first.

markusweimer commented:
> @jwang98052 may want to close this PR first.

+1 on only discussing one of these related pull requests at a time. Many comments are redundant between the two. Let's get one in first, and then do the second state machine based on that design.

jwang98052 commented:
Close it and open a new one after #948 is merged.
